File: Backend/src/test/java/lan/dk/podcastserver/manager/worker/selector/UpdaterSelectorTest.java
package lan.dk.podcastserver.manager.worker.selector;

import io.vavr.collection.HashSet;
import io.vavr.collection.Set;
import lan.dk.podcastserver.manager.worker.updater.*;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class UpdaterSelectorTest {

    private UpdaterSelector updaterSelector;

    private @Mock BeInSportsUpdater beInSportsUpdater;
    private @Mock CanalPlusUpdater canalPlusUpdater;
    private @Mock DailymotionUpdater dailymotionUpdater;
    private @Mock GulliUpdater gulliUpdater;
    private @Mock JeuxVideoComUpdater jeuxVideoComUpdater;
    private @Mock FranceTvUpdater franceTvUpdater;
    private @Mock RSSUpdater rssUpdater;
    private @Mock SixPlayUpdater sixPlayUpdater;
    private @Mock TF1ReplayUpdater tf1ReplayUpdater;
    private @Mock UploadUpdater uploadUpdater;
    private @Mock YoutubeUpdater youtubeUpdater;

    @Before
    public void setUp() throws Exception {
        /* Given */
        HashSet<Updater> updaters = HashSet.of(
                beInSportsUpdater, canalPlusUpdater, dailymotionUpdater, gulliUpdater,
                jeuxVideoComUpdater, franceTvUpdater, rssUpdater, sixPlayUpdater,
                tf1ReplayUpdater, uploadUpdater, youtubeUpdater
        );

        updaters.forEach(u -> {
            when(u.compatibility(anyString())).thenCallRealMethod();
            when(u.type()).thenCallRealMethod();
        });

        updaterSelector = new UpdaterSelector(updaters.toJavaSet());
    }


    @Test
    public void should_return_a_BeInSportUpdater () {
        /* When */ Updater updaterClass = updaterSelector.of("http://www.beinsports.com/replay/category/3361/name/lexpresso");
        /* Then */ assertThat(updaterClass).isEqualTo(beInSportsUpdater);
    }

    @Test
    public void should_return_a_CanalPlusUpdater() {
        /* When */ Updater updaterClass = updaterSelector.of("http://www.canalplus.fr/show/for/dummies");
        /* Then */ assertThat(updaterClass).isEqualTo(canalPlusUpdater);
    }

    @Test
    public void should_return_a_DailymotionUpdater() {
        /* When */ Updater updaterClass = updaterSelector.of("http://www.dailymotion.com/showname");
        /* Then */ assertThat(updaterClass).isEqualTo(dailymotionUpdater);
    }

    @Test
    public void should_return_a_GulliUpdater() {
        /* When */ Updater updaterClass = updaterSelector.of("http://replay.gulli.fr/showname");
        /* Then */ assertThat(updaterClass).isEqualTo(gulliUpdater);
    }

    @Test
    public void should_return_a_JeuxVideoComUpdater() {
        /* When */ Updater updaterClass = updaterSelector.of("http://www.jeuxvideo.com/show/for/dummies");
        /* Then */ assertThat(updaterClass).isEqualTo(jeuxVideoComUpdater);
    }

    @Test
    public void should_return_a_FranceTvUpdate() {
        /* When */ Updater updaterClass = updaterSelector.of("http://www.france.tv/show/for/dummies");
        /* Then */ assertThat(updaterClass).isEqualTo(franceTvUpdater);
    }

    @Test
    public void should_return_an_RssUpdater () {
        /* When */ Updater updaterClass = updaterSelector.of("www.link.to.rss/feeds");
        /* Then */ assertThat(updaterClass).isEqualTo(rssUpdater);
    }

    @Test
    public void should_return_an_SixPlayUpdater () {
        /* When */ Updater updaterClass = updaterSelector.of("http://www.6play.fr/turbo_test");
        /* Then */ assertThat(updaterClass).isEqualTo(sixPlayUpdater);
    }

    @Test
    public void should_return_an_Tf1ReplayUpdater () {
        /* When */ Updater updaterClass = updaterSelector.of("http://www.tf1.fr/title");
        /* Then */ assertThat(updaterClass).isEqualTo(tf1ReplayUpdater);
    }

    @Test
    public void should_return_a_YoutubeUpdater () {
        /* When */ Updater updaterClass = updaterSelector.of("http://www.youtube.com/user/fakeUser");
        /* Then */ assertThat(updaterClass).isEqualTo(youtubeUpdater);
    }

    @Test
    public void should_reject_empty_url() {
        /* When */
        assertThat(updaterSelector.of("")).isEqualTo(UpdaterSelector.NO_OP_UPDATER);
    }

    @Test
    public void should_serve_types() {
        /* When */
        Set<AbstractUpdater.Type> types = updaterSelector.types();

        /* Then */
        assertThat(types).isNotEmpty().hasSize(11);
    }

}

File: backend/src/test/kotlin/com/github/davinkevin/podcastserver/update/updaters/UpdaterSelectorTest.kt
package com.github.davinkevin.podcastserver.update.updaters

import com.github.davinkevin.podcastserver.update.updaters.dailymotion.DailymotionUpdater
import com.github.davinkevin.podcastserver.update.updaters.francetv.FranceTvUpdater
import com.github.davinkevin.podcastserver.update.updaters.gulli.GulliUpdater
import com.github.davinkevin.podcastserver.update.updaters.mytf1.MyTf1Updater
import com.github.davinkevin.podcastserver.update.updaters.rss.RSSUpdater
import com.github.davinkevin.podcastserver.update.updaters.upload.UploadUpdater
import com.github.davinkevin.podcastserver.update.updaters.youtube.YoutubeByApiUpdater
import com.github.davinkevin.podcastserver.update.updaters.youtube.YoutubeByXmlUpdater
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.junit.jupiter.MockitoSettings
import org.mockito.kotlin.any
import org.mockito.kotlin.whenever
import org.mockito.quality.Strictness
import java.net.URI
import java.util.stream.Stream

@ExtendWith(MockitoExtension::class)
@MockitoSettings(strictness = Strictness.LENIENT)
class UpdaterSelectorTest {

    private lateinit var  updaterSelector: UpdaterSelector

    @Mock lateinit var dailymotionUpdater: DailymotionUpdater
    @Mock lateinit var franceTvUpdater: FranceTvUpdater
    @Mock lateinit var gulliUpdater: GulliUpdater
    @Mock lateinit var rssUpdater: RSSUpdater
    @Mock lateinit var myTf1Updater: MyTf1Updater
    @Mock lateinit var uploadUpdater: UploadUpdater
    @Mock lateinit var youtubeByXmlUpdater: YoutubeByXmlUpdater
    @Mock lateinit var youtubeByApiUpdater: YoutubeByApiUpdater

    @BeforeEach
    fun beforeEach() {
        /* Given */
        val updaters = setOf(
                dailymotionUpdater,
                franceTvUpdater, 
                gulliUpdater,
                rssUpdater,
                myTf1Updater,
                uploadUpdater, 
                youtubeByXmlUpdater,
                youtubeByApiUpdater
        )

        updaters.forEach {
            whenever(it.compatibility(any())).thenCallRealMethod()
            whenever(it.type()).thenCallRealMethod()
        }

        updaterSelector = UpdaterSelector(updaters)
    }

    @MethodSource("urlToUpdater")
    @DisplayName("should return")
    @ParameterizedTest(name = "{1} updater for {0}")
    fun `should return matching updater`(url: String, type: String) {
        /* When */
        val updaterClass = updaterSelector.of(URI(url))
        /* Then */
        assertThat(updaterClass.type().key).isEqualTo(type)
    }

    @Test
    fun `should serve types`() {
        /* Given */
        val uTypes: List<Type> = setOf(
                dailymotionUpdater,
                franceTvUpdater,
                gulliUpdater,
                rssUpdater,
                myTf1Updater,
                youtubeByXmlUpdater,
                youtubeByApiUpdater
        )
                .map { it.type() }


        /* When */
        val types = updaterSelector.types()

        /* Then */
        assertThat(types)
                .containsAll(uTypes)
                .hasSize(7)
    }

    companion object {
        @JvmStatic
        fun urlToUpdater() =
                Stream.of(
                        Arguments.of("http://www.dailymotion.com/showname", "Dailymotion"),
                        Arguments.of("http://www.france.tv/show/for/dummies", "FranceTv"),
                        Arguments.of("http://replay.gulli.fr/showname", "Gulli"),
                        Arguments.of("http://www.link.to.rss/feeds", "RSS"),
                        Arguments.of("http://www.tf1.fr/title", "MyTF1"),
                        Arguments.of("http://www.youtube.com/user/fakeUser", "Youtube")
                )
    }
}
